ray-curator/ray_curator/stages/deduplication/semantic/pairwise.py

```diff
@@ -0,0 +1,326 @@
+            )
+            return FileGroupTask(
+                task_id=task.task_id,
+                dataset_name=task.dataset_name,
+                _metadata={
+                    **task._metadata,
+                    "centroid_id": cluster_id,
+                },
+                _stage_perf=task._stage_perf,
+                data=[os.path.join(self.output_path, f"cluster_{cluster_id}.parquet")],
+            )
+
+        # Cannot concatenate dataframes with embeddings due to cudf 2bn row limit
+        # Instead, concatenate metadata columns and handle embeddings separately
+        metadata_dfs, embedding_arrays = [], []
+        for df in dfs:
+            metadata_dfs.append(df[metadata_cols])
+            embedding_arrays.append(get_array_from_df(df, self.embedding_field))
+
+        metadata_cluster_df = cudf.concat(metadata_dfs, ignore_index=True).reset_index(drop=True)
+
+        # Add original index to track reordering
+        metadata_cluster_df["_original_idx"] = metadata_cluster_df.index
+
+        ranked_metadata_df = self.ranking_strategy.rank_cluster(metadata_cluster_df)
+        # Get reorder indices from the ranked dataframe (TODO: we get it to CPU, but maybe we can do it on GPU todo)
+        reorder_indices = ranked_metadata_df["_original_idx"].to_arrow().to_pylist()
+        # Remove the helper column
+        ranked_metadata_df = ranked_metadata_df.drop(columns=["_original_idx"])
+
+        # Convert numpy arrays to torch tensors before concatenating
+        concatenated_embeddings = torch.cat([torch.as_tensor(arr, device="cuda") for arr in embedding_arrays], dim=0)
+        cluster_embeddings = concatenated_embeddings[reorder_indices]
+
+        ids = ranked_metadata_df[self.id_field]
+
+        # Compute pairwise similarities
+        max_similarity, max_indices = pairwise_cosine_similarity_batched(
+            cluster_embeddings, "cuda", self.pairwise_batch_size
+        )
+
+        # Convert indices back to IDs
+        max_indices_id = ids.iloc[max_indices].reset_index(drop=True)
+
+        # Create result dataframe
+        points_to_remove_df = cudf.DataFrame(
+            {
+                **({CURATOR_DEDUP_ID_STR: ids} if has_curator_id else {}),
+                "id": ids,
+                "max_id": max_indices_id,
+                "cosine_sim_score": max_similarity,
+            }
+        )
+
+        # Write results
+        self.write_parquet(points_to_remove_df, output_path, storage_options=self.output_storage_options, index=False)
+
+        t2 = time.perf_counter()
+        if self.verbose:
+            logger.debug(
+                f"Pairwise computation for cluster {cluster_id} with {num_rows} rows done in {(t2 - t1):.2f} seconds"
+            )
+
+        return FileGroupTask(
+            task_id=task.task_id,
+            dataset_name=task.dataset_name,
+            _metadata={**task._metadata, "centroid_id": cluster_id},
+            _stage_perf=task._stage_perf,
```

Comment:
Pass storage options here too

Reply:
Discussed offline don't think Task should have storage_options